lib/byron/exe/cardano-wallet-byron.hs

```diff
@@ -220,18 +245,68 @@ cmdServe = command "serve" $ info (helper <*> helper' <*> cmd) $ mempty
                                  Options
 -------------------------------------------------------------------------------}
 
+data NetworkConfiguration where
+    MainnetConfig
+        :: (SomeNetworkDiscriminant, NodeVersionData)
+        -> NetworkConfiguration
+
+    TestnetConfig
+        :: (SomeNetworkDiscriminant, NodeVersionData)
+        -> FilePath
+        -> NetworkConfiguration
+
+-- | Hand-written as there's no Show instance for 'NodeVersionData'
+instance Show NetworkConfiguration where
+    show = \case
+        MainnetConfig{} ->
+            "MainnetConfig"
+        TestnetConfig _ genesisFile ->
+            "TestnetConfig " <> show genesisFile
+
 -- | --node-socket=FILE
 nodeSocketOption :: Parser FilePath
 nodeSocketOption = optionT $ mempty
     <> long "node-socket"
     <> metavar "FILE"
     <> help "Path to the node's domain socket."
 
--- | --testnet|--mainnet
-networkDiscriminantFlag :: Parser SomeNetworkDiscriminant
-networkDiscriminantFlag =
-        flag' (SomeNetworkDiscriminant $ Proxy @'Mainnet) (long "mainnet")
-    <|> flag' (SomeNetworkDiscriminant $ Proxy @'Testnet) (long "testnet")
+-- | --mainnet | (--testnet=NETWORK_MAGIC --genesis=FILE)
+networkConfigurationOption :: Parser NetworkConfiguration
+networkConfigurationOption =
+    mainnetFlag <|> (TestnetConfig <$> testnetOption <*> genesisOption)
+  where
+    -- --mainnet
+    mainnetFlag = flag'
+        (MainnetConfig (SomeNetworkDiscriminant $ Proxy @'Mainnet, mainnetVersionData))
+        (long "mainnet")
+
+    -- --testnet NETWORK_MAGIC
+    testnetOption = fmap someTestnetDiscriminant $ optionT $ mempty
+        <> long "testnet"
+        <> metavar "NETWORK_MAGIC"
+      where
+        someTestnetDiscriminant
+            :: ProtocolMagic
+            -> (SomeNetworkDiscriminant, NodeVersionData)
+        someTestnetDiscriminant pm@(ProtocolMagic n) =
+            case someNatVal (fromIntegral n) of
+                Just (SomeNat proxy) ->
+                    ( SomeNetworkDiscriminant $ mapProxy proxy
+                    , testnetVersionData pm
+                    )
+                _ -> error "networkDiscriminantFlag: impossible: failed to \
```

Comment:
maybe `impossible` was in the earlier version and forgot to erase? `failed to convert ProtocolMagic to SomeNat` is not enough?

Reply:
fixed.